File: server/src/main/java/org/e8yes/sql/SqlRunner.java
/**
 * e8yes demo web server.
 *
 * <p>Copyright (C) 2020 Yichen Ma {yichenm2@uci.edu}, Chifeng Wen {daviesx66@gmail.com}
 *
 * <p>This program is free software: you can redistribute it and/or modify it under the terms of the
 * GNU General Public License as published by the Free Software Foundation, either version 3 of the
 * License, or (at your option) any later version.
 *
 * <p>This program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY;
 * without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU General Public License for more details.
 *
 * <p>You should have received a copy of the GNU General Public License along with this program. If
 * not, see <http://www.gnu.org/licenses/>.
 */
package org.e8yes.sql;

import java.lang.reflect.InvocationTargetException;
import java.sql.SQLException;
import java.time.Instant;
import java.util.List;
import java.util.Set;
import java.util.TreeSet;
import org.e8yes.sql.connection.ConnectionInterface;
import org.e8yes.sql.connection.ConnectionReservoirInterface;
import org.e8yes.sql.orm.DataCollection;
import org.e8yes.sql.orm.QueryCompletion;
import org.e8yes.sql.primitive.SqlInt;
import org.e8yes.sql.primitive.SqlLong;
import org.e8yes.sql.primitive.SqlStr;
import org.e8yes.sql.resultset.ResultSetInterface;

/** Collects information from different sources to build then execute an SQL query. */
public class SqlRunner {

  /**
   * Constructs and runs query with partial information provided and synchronously returns the
   * result.
   *
   * @param <ReturnType>
   * @param query Partial query. Example: "AUser user JOIN CreditCard card ON card.userId = user.id
   *     WHERE user.joinDate > '2020-1-1'"
   * @param entityType The type of Java object to run ORM on. It will convert query result to this
   *     target entity type
   * @param reservoir Connection reservoir to allocate database connections.
   * @return The query result.
   * @throws NoSuchMethodException
   * @throws InstantiationException
   * @throws IllegalAccessException
   * @throws IllegalArgumentException
   * @throws InvocationTargetException
   * @throws SQLException
   */
  public static <ReturnType> List<ReturnType> runQuery(
      SqlQueryBuilder query, Class entityType, ConnectionReservoirInterface reservoir)
      throws SQLException, NoSuchMethodException, InstantiationException, IllegalAccessException,
          IllegalArgumentException, InvocationTargetException {
    String completedQuery = QueryCompletion.completeSelectQuery(query.jdbcQuery(), entityType);
    ConnectionInterface conn = reservoir.take();
    ResultSetInterface rs = conn.runQuery(completedQuery, query.queryParams());
    List<ReturnType> results = DataCollection.collect(rs, entityType);
    reservoir.put(conn);
    return results;
  }

  /**
   * Constructs and runs query then checks if the query returns at least 1 result.
   *
   * @param query Partial query to test against. Example: "AUser user JOIN CreditCard card ON
   *     card.userId = user.id WHERE user.joinDate > '2020-1-1'"
   * @param reservoir Connection reservoir to allocate database connections.
   * @return true if at least 1 entry returns from the query, otherwise, false.
   * @throws SQLException
   */
  public static boolean runExists(SqlQueryBuilder query, ConnectionReservoirInterface reservoir)
      throws SQLException {
    String completedQuery = "SELECT TRUE FROM " + query.jdbcQuery();
    ConnectionInterface conn = reservoir.take();
    ResultSetInterface rs = conn.runQuery(completedQuery, query.queryParams());
    boolean exists = rs.hasNext();
    reservoir.put(conn);

    return exists;
  }

  /**
   * Runs a deletion SQL query on the specified table.
   *
   * @param <ReturnType>
   * @param tableName Name of the table to run deletion query on.
   * @param query Partial deletion query. Example: "WHERE id = 100"
   * @param reservoir Connection reservoir to allocate database connections.
   * @return The number of rows deleted.
   * @throws SQLException
   */
  public static <ReturnType> int runDelete(
      String tableName, SqlQueryBuilder query, ConnectionReservoirInterface reservoir)
      throws SQLException {
    String completedQuery = "DELETE FROM " + tableName + " " + query.jdbcQuery();
    ConnectionInterface conn = reservoir.take();
    int numRowsUpdated = conn.runUpdate(completedQuery, query.queryParams());
    reservoir.put(conn);

    return numRowsUpdated;
  }

  /**
   * Save a Java object to the specified SQL table.
   *
   * @param <Type>
   * @param record Java object to be saved.
   * @param tableName Target SQL table to save to.
   * @param entityType The type of Java object to run ORM on. It will convert the entity type to the
   *     SQL row format.
   * @param override Whether or not to override existing record when conflict occurs on update.
   * @param reservoir Connection reservoir to allocate database connections.
   * @return The number of SQL rows affected by this update.
   * @throws SQLException
   * @throws IllegalAccessException
   */
  public static <Type> int runUpdate(
      Type record,
      String tableName,
      Class entityType,
      boolean override,
      ConnectionReservoirInterface reservoir)
      throws SQLException, IllegalAccessException {
    String completedQuery = QueryCompletion.completeInsertQuery(tableName, entityType, override);
    ConnectionInterface.QueryParams params =
        QueryCompletion.generateInsertQueryParams(record, entityType, override);
    ConnectionInterface conn = reservoir.take();
    int numRowsUpdated = conn.runUpdate(completedQuery, params);
    reservoir.put(conn);

    return numRowsUpdated;
  }

  /**
   * Get all the database tables.
   *
   * @param reservoir Connection reservoir to allocate database connections.
   * @return table name of all the database tables.
   * @throws java.sql.SQLException
   */
  public static Set<String> tables(ConnectionReservoirInterface reservoir) throws SQLException {
    ConnectionInterface conn = reservoir.take();
    String reflectionQuery =
        "SELECT tb.table_name FROM information_schema.tables tb "
            + "WHERE tb.table_schema='public'";
    ResultSetInterface rs = conn.runQuery(reflectionQuery, new ConnectionInterface.QueryParams());

    Set<String> tableNames = new TreeSet();
    SqlStr tableName = new SqlStr();
    for (; rs.hasNext(); rs.next()) {
      rs.setCellValueToField(1, tableName);
      assert (tableName.value() != null);
      tableNames.add(tableName.value());
    }

    reservoir.put(conn);

    return tableNames;
  }

  /**
   * Send a heartbeat to test the connection.
   *
   * @param reservoir Connection reservoir to allocate database connections.
   * @throws SQLException
   */
  public static void sendHeartBeat(ConnectionReservoirInterface reservoir) throws SQLException {
    ConnectionInterface conn = reservoir.take();
    ResultSetInterface rs = conn.runQuery("SELECT 1", new ConnectionInterface.QueryParams());

    SqlInt one = new SqlInt();
    rs.setCellValueToField(1, one);
    assert (one.value() != null);
    assert (one.value() == 1);

    reservoir.put(conn);
  }

  /**
   * Generate approximately unique ID from time.
   *
   * @return unique ID.
   */
  public static long timeId() {
    Instant inst = Instant.now();
    long micros = inst.getEpochSecond() * 1000 * 1000 + inst.getNano() / 1000;
    return Long.reverse(micros);
  }

  /**
   * Generate sequential ID from a sequence table.
   *
   * @param seqTable Name of the sequence table to generate ID from.
   * @param reservoir Connection reservoir to allocate database connections.
   * @return unique ID.
   * @throws SQLException
   */
  public static long seqId(String seqTable, ConnectionReservoirInterface reservoir)
      throws SQLException {
    ConnectionInterface conn = reservoir.take();

    ResultSetInterface rs =
        conn.runQuery("SELECT nextval('" + seqTable + "')", new ConnectionInterface.QueryParams());
    assert (rs.hasNext());
    SqlLong id = new SqlLong();
    rs.setCellValueToField(0, id);
    assert (id.value() != null);

    reservoir.put(conn);

    return id.value();
  }
}

File: demoweb_service/demoweb/sql/sql_runner.cc
/**
 * e8yes demo web server.
 *
 * <p>Copyright (C) 2020 Chifeng Wen {daviesx66@gmail.com}
 *
 * <p>This program is free software: you can redistribute it and/or modify it under the terms of the
 * GNU General Public License as published by the Free Software Foundation, either version 3 of the
 * License, or (at your option) any later version.
 *
 * <p>This program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY;
 * without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU General Public License for more details.
 *
 * <p>You should have received a copy of the GNU General Public License along with this program. If
 * not, see <http://www.gnu.org/licenses/>.
 */

#include <chrono>
#include <cstdint>
#include <exception>
#include <ios>
#include <memory>
#include <string>
#include <unordered_set>

#include "sql/connection/connection_interface.h"
#include "sql/connection/connection_reservoir_interface.h"
#include "sql/orm/query_completion.h"
#include "sql/reflection/sql_primitives.h"
#include "sql/resultset/result_set_interface.h"
#include "sql/sql_runner.h"

namespace e8 {
namespace {

int64_t reverse_bytes(int64_t original) {
    return (original & 0xFF) << 56 | ((original >> 8) & 0xFF) << 48 |
           ((original >> 16) & 0xFF) << 40 | ((original >> 24) & 0xFF) << 32 |
           ((original >> 32) & 0xFF) << 24 | ((original >> 40) & 0xFF) << 16 |
           ((original >> 48) & 0xFF) << 8 | ((original >> 56) & 0xFF);
}

} // namespace

uint64_t Update(SqlEntityInterface const &entity, std::string const &table_name, bool override,
                ConnectionReservoirInterface *reservoir) {
    InsertQueryAndParams query_and_params = GenerateInsertQuery(table_name, entity, override);

    ConnectionInterface *conn = reservoir->Take();
    uint64_t numRowsUpdated =
        conn->run_update(query_and_params.query, query_and_params.query_params);
    reservoir->Put(conn);

    return numRowsUpdated;
}

uint64_t Delete(std::string const &table_name, SqlQueryBuilder const &query,
                ConnectionReservoirInterface *reservoir) {
    std::string completed_query = "DELETE FROM " + table_name + " " + query.psql_query();

    ConnectionInterface *conn = reservoir->Take();
    uint64_t numRowsUpdated = conn->run_update(completed_query, query.query_params());
    reservoir->Put(conn);

    return numRowsUpdated;
}

bool Exists(SqlQueryBuilder const &query, ConnectionReservoirInterface *reservoir) {
    std::string exists_query = "SELECT TRUE FROM " + query.psql_query();

    ConnectionInterface *conn = reservoir->Take();

    std::unique_ptr<ResultSetInterface> rs = conn->run_query(exists_query, query.query_params());
    bool exists = rs->has_next();

    reservoir->Put(conn);

    return exists;
}

std::unordered_set<std::string> Tables(ConnectionReservoirInterface *reservoir) {
    ConnectionInterface *conn = reservoir->Take();
    std::string reflection_query =
        "SELECT tb.table_name FROM information_schema.tables tb WHERE tb.table_schema='public'";
    std::unique_ptr<ResultSetInterface> rs =
        conn->run_query(reflection_query, ConnectionInterface::QueryParams());

    std::unordered_set<std::string> table_names;
    SqlStr table_name("table_name");
    for (; rs->has_next(); rs->next()) {
        rs->set_field(0, &table_name);
        assert(table_name.value().has_value());
        table_names.insert(table_name.value().value());
    }

    reservoir->Put(conn);

    return table_names;
}

void SendHeartBeat(ConnectionReservoirInterface *reservoir) {
    ConnectionInterface *conn = reservoir->Take();

    std::unique_ptr<ResultSetInterface> rs =
        conn->run_query("SELECT 1", ConnectionInterface::QueryParams());

    if (!rs->has_next()) {
        throw std::ios_base::failure("Heart beat results in an empty result set.");
    }

    SqlInt one("heart_beat");
    rs->set_field(0, &one);
    if (!one.value().has_value() || one.value().value() != 1) {
        throw std::system_error();
    }

    reservoir->Put(conn);
}

int64_t TimeId() {
    auto now = std::chrono::high_resolution_clock::now();
    auto nanos = std::chrono::time_point_cast<std::chrono::nanoseconds>(now);
    auto dura = std::chrono::duration_cast<std::chrono::nanoseconds>(nanos.time_since_epoch());
    int64_t timestamp = dura.count();
    return reverse_bytes(timestamp);
}

int64_t SeqId(std::string const &seq_table, ConnectionReservoirInterface *reservoir) {
    ConnectionInterface *conn = reservoir->Take();

    std::unique_ptr<ResultSetInterface> rs =
        conn->run_query("SELECT nextval('" + seq_table + "')", ConnectionInterface::QueryParams());
    assert(rs->has_next());

    SqlLong id("id");
    rs->set_field(0, &id);
    assert(id.value().has_value());

    reservoir->Put(conn);

    return id.value().value();
}

} // namespace e8
